docs/source/developer_guide.md

```diff
@@ -260,6 +260,97 @@ Sometimes, we need to temporarily change the log pattern (eg: for reporting deci
+
+The dispatching by `expensive.cuh` is performed as follows:
+``` c++
+#if !defined(RAFT_EXPLICIT_INSTANTIATE_ONLY)
+// If implicit instantiation is allowed, include template definitions.
+#include "expensive-inl.cuh"
+#endif
+
+#ifdef RAFT_COMPILED
+// Include extern template instantiations when RAFT is compiled.
+#include "expensive-ext.cuh"
+#endif
+```
+
+The file `expensive-inl.cuh` is unchanged:
+``` c++
+namespace raft {
+template <typename T>
+void expensive(T arg) {
+  // .. function body
+}
+} // namespace raft
+```
+
+The file `expensive-ext.cuh` contains the following:
+``` c++
+#include <raft/util/raft_explicit.cuh> // RAFT_EXPLICIT
+
+#ifdef RAFT_EXPLICIT_INSTANTIATE_ONLY
+namespace raft {
+// (1) define templates to raise an error in case of accidental instantiation 
+template <typename T> void expensive(T arg) RAFT_EXPLICIT;
+} // namespace raft
+#endif //RAFT_EXPLICIT_INSTANTIATE_ONLY
+
+// (2) Provide extern template instantiations.
+extern template void raft::expensive<int>(int);
+extern template void raft::expensive<float>(float);
+```
+
+This header has two responsibilities: (1) define templates to raise an error in case of accidental instantiation and (2) provide `extern template` instantiations.
+First, if `RAFT_EXPLICIT_INSTANTIATE_ONLY` is set, `expensive` is defined. This is done for two reasons: (1) to give a definition, because the definition in `expensive-inl.cuh` was skipped and (2) to indicate that the template should be explicitly instantiated by taging it with the `RAFT_EXPLICIT` macro. This macro defines the function body, and it ensures that an informative error message is generated when an implicit instantiation erroneously occurs. Finally, the `extern template` instantiations are listed.
+
+To actually generate the code for the template instances, the file `src/expensive.cu` contains the following. Note that the only difference between the extern template instantiations in `expensive-ext.cuh` and these lines are the removal of the word `extern`:
+
+``` c++
+#include <raft/expensive-inl.cuh>
+
+template void raft::expensive<int>(int);
+template void raft::expensive<float>(float);
+```
+
+**Design considerations**: 
+
+1. In the `-ext.cuh` header, do not include implementation headers. Only include function parameter types and types that are used to instantiate the templates. If a primitive takes custom parameter types, define them in a separate header called `<primitive_name>_types.hpp`. 
```

Comment:
```suggestion
1. In the `-ext.cuh` header, do not include implementation headers. Only include function parameter types and types that are used to instantiate the templates. If a primitive takes custom parameter types, define them in a separate header called `<primitive_name>_types.hpp` (see [Common Design Considerations](https://github.com/rapidsai/raft/blob/7b065aff81a0b1976e2a9e2f3de6690361a1111b/docs/source/developer_guide.md#common-design-considerations)).  
```

Reply:
Do you happen to know how we can make the link always resolve to our latest branch? Now it is pointing to this specific commit. I could point it to the 23.06 branch, but that will be outdated in a few months too...